gax-httpjson/src/test/java/com/google/api/gax/httpjson/ApiMessageOperationTransformersTest.java

```diff
@@ -0,0 +1,228 @@
+
+    Truth.assertThat(transformer.apply(operationSnapshot)).isEqualTo(emptyResponse);
+  }
+
+  @Test
+  public void testResponseTransformer_exception() {
+    thrown.expect(UnavailableException.class);
+    ResponseTransformer<EmptyMessage> transformer = ResponseTransformer.create(EmptyMessage.class);
+    EmptyMessage emptyResponse = new EmptyMessage();
+    MetadataMessage metadata = new MetadataMessage(Status.PENDING, Code.UNAVAILABLE);
+    OperationSnapshot operationSnapshot =
+        new OperationSnapshotImpl(
+            new OperationMessage<>("Unavailable; no response method", emptyResponse, metadata));
+
+    Truth.assertThat(transformer.apply(operationSnapshot)).isEqualTo(emptyResponse);
+  }
+
+  @Test
+  public void testResponseTransformer_mismatchedTypes() {
+    thrown.expect(ApiException.class);
+    thrown.expectMessage("cannot be cast");
+    ResponseTransformer<EmptyMessage> transformer = ResponseTransformer.create(EmptyMessage.class);
+    MetadataMessage metadata = new MetadataMessage(Status.PENDING, Code.OK);
+    ApiMessage bananaResponse =
+        new FakeApiMessage(ImmutableMap.<String, Object>of("name", "banana"), null, null);
+    EmptyMessage emptyResponse = new EmptyMessage();
+    OperationSnapshot operationSnapshot =
+        new OperationSnapshotImpl(
+            new OperationMessage<>("No response method", bananaResponse, metadata));
+    Truth.assertThat(transformer.apply(operationSnapshot)).isEqualTo(emptyResponse);
+  }
+
+  @Test
+  public void testMetadataTransformer() {
+    MetadataTransformer<MetadataMessage> transformer =
+        MetadataTransformer.create(MetadataMessage.class);
+    EmptyMessage returnType = new EmptyMessage();
+    MetadataMessage metadataMessage = new MetadataMessage(Status.PENDING, Code.OK);
+    OperationMessage operation = new OperationMessage<>("foo", returnType, metadataMessage);
+    OperationSnapshot operationSnapshot = new OperationSnapshotImpl(operation);
+    Truth.assertThat(transformer.apply(operationSnapshot)).isEqualTo(metadataMessage);
+  }
+
+  @Test
+  public void testMetadataTransformer_mismatchedTypes() {
+    thrown.expect(ApiException.class);
+    thrown.expectMessage("cannot be cast");
+    MetadataTransformer<OperationMessage> transformer =
+        MetadataTransformer.create(OperationMessage.class);
+    MetadataMessage metadataMessage = new MetadataMessage(Status.PENDING, Code.OK);
+    ApiMessage bananaResponse =
+        new FakeApiMessage(ImmutableMap.<String, Object>of("name", "banana"), null, null);
+    OperationMessage metadata =
+        new OperationMessage<>("No response method", bananaResponse, metadataMessage);
+    OperationSnapshot operationSnapshot = new OperationSnapshotImpl(metadata);
+    Truth.assertThat(transformer.apply(operationSnapshot)).isEqualTo(bananaResponse);
+  }
+
+  private enum Status {
+    PENDING,
+    DONE
+  }
+
+  private static class MetadataMessage<ResponseT extends ApiMessage> implements ApiMessage {
```

Comment:
Can these classes be replaced with `Mockito` mocks (we should avoid creating mocks manually where possible, because they are usually much longer and require lots of boilerplate code, which has to be updated if interface changes (here the `ApiMessage`)). If it is not possible to use mocks instead, please  rename them to `Fake*` or `Mock*` or something like that, so they do not appear as false positive results in code search (when searching for real stuff).

Reply:
Naming them `Fake*`. Because the test logic relies on calling class type methods like `instanceof`, I wasn't sure how to mock that with mockito.